auth/api/iam/error.go

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package iam
+
+import (
+	"errors"
+	"github.com/labstack/echo/v4"
+	"github.com/nuts-foundation/nuts-node/core"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// ErrorCode specifies error codes as defined by the OAuth2 specifications.
+// Codes and descriptions are taken from https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1
+type ErrorCode string
+
+const (
+	// InvalidRequest is returned when the request is missing a required parameter, includes an invalid parameter value,
+	// includes a parameter more than once, or is otherwise malformed.
+	InvalidRequest ErrorCode = "invalid_request"
+	// UnsupportedResponseType is returned when the authorization server does not support obtaining an authorization code using this method.
+	UnsupportedResponseType ErrorCode = "unsupported_response_type"
+	// ServerError is returned when the Authorization Server encounters an unexpected condition that prevents it from fulfilling the request.
+	ServerError ErrorCode = "server_error"
+)
+
+// Make sure the error implements core.HTTPStatusCodeError, so the HTTP request logger can log the correct status code.
+var _ core.HTTPStatusCodeError = OAuth2Error{}
+
+// OAuth2Error is an OAuth2 error that signals the error was (probably) caused by the client (e.g. bad request),
+// or that the client can recover from the error (e.g. retry).
+type OAuth2Error struct {
```

Comment:
There are also some generated error structs, those can be alliased?

Reply:
Defined as:

```
// ErrorResponse defines model for ErrorResponse.
type ErrorResponse struct {
	// Error Code identifying the error that occurred.
	Error string `json:"error"`
}
```

Our error type also has some helper fields (RedirectURI, InternalError) and additional functions, so I don't see much value in using the generated type.

Reply:
Not using the generated type, but reusing this new type and omit the generated type from generation.